text/1015-full-text-search.rst

```diff
@@ -58,74 +58,272 @@ Specification
+single object type. So at most one FTS index can be defined on any type
+(subject to ``??`` once the extension enabling/disabling is implemented). In
+particular this means that when an FTS index is inherited by a type and a new
+index definition is provided in the extending type, the new definition
+completely overrides the inherited one.
+
+For example::
+
+  type Document {
+      body: str;
+      index fts::textsearch on (fts::language(.body, 'english'));
+  }
+
+  type InternalDoc extending Document {
+      internal: str;
+  }
+
+  type TitledDoc extending Document {
+      title: str;
+      index fts::textsearch on (fts::language(.title, 'english'));
+  }
+
+The ``InternalDoc`` type has no FTS index of its own, so it inherits the index
+from ``Document``. Thus only the ``.body`` is indexed. Conversely,
+``TitledDoc`` defines a new ``fts::textsearch`` index on ``.title``, but the
+``body`` property would no longer be indexed for ``TitledDoc`` objects.
+
+The reason for this design is that we're limited to having no more than one
+FTS index per type as it covers all search functionality. This means that we
+must have a way of overriding an existing index with type-specific one. The
+least surprising way of doing that is for the explicitly specified index to
+override the inherited one.
+
+In case of multiple inheritance indexes may conflict with one another, that is
+considered an error and the user is required to explicitly provide an FTS
+index to resolve this (to minimize ambiguity).
+
+All of these extra rules apply only to FTS indexes and not indexes in general.
+
+
+Search Functions
+----------------
+
+An important feature of the search function is that it actually performs the
+job typically done by the ``FILTER`` clause. The function is supposed to
+filter a bunch of objects based on the FTS query and indexed fields. In
+addition to filtering, the results may be annotated with relevance (ranked)
+and potentially with highlighted matches::
+
+  function std::fts::search(
+      doc: anyobject,
+      query: str,
+      named only language: str,
+      named only weights: optional array<float64> = {},
+      named only rank_opts: optional str = 'default',
+      named only highlight_opts: optional str = {}
+  ) -> optional tuple<
+    object: anyobject,
+    rank: float64,
+    highlights: array<str>
+  >
+
+The ``doc`` input provides the object that should be searched. The
+details of which properties should be searched and other FTS parameters will
+be provided by the FTS index on the specified type. Searching an unindexed
+type should simply produce no matches (resulting in an ``{}``).
+
+The ``language`` parameter indicated which language the query is using and
+therefore allows to target only the relevant documents if there are multiple
+languages.
+
+If ranking was requested (which is the default behavior), then the value will
+be written in ``rank``, otherwise the value ``0`` is used.
```

Comment:
Rank will always be requested, right?

It might not be used in the result of the query, in which case the compiler can make sure it is not even computed, but it will never default to 0.